File: src/381.o-1-时间插入、删除和获取随机元素-允许重复.py
#
# @lc app=leetcode.cn id=381 lang=python3
#
# [381] O(1) 时间插入、删除和获取随机元素 - 允许重复
#

import unittest
# @lc code=start
import random
from collections import defaultdict
from typing import Dict, Set


class RandomizedCollection:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.items = []
        self.indexes: Dict[int, Set] = defaultdict(set)

    def insert(self, val: int) -> bool:
        """
        Inserts a value to the collection. Returns true if the collection did not already contain the specified element.
        """
        # print(f"insert {val}")
        already = val in self.indexes
        self.items.append(val)
        self.indexes[val].add(len(self.items) - 1)
        return not already

    def remove(self, val: int) -> bool:
        """
        Removes a value from the collection. Returns true if the collection contained the specified element.
        """
        # print(f"remove {val}, indexes={self.indexes}, items={self.items}")
        if val not in self.indexes:
            return False
        n = len(self.items)

        if self.items[-1] == val:
            self.indexes[val].remove(n-1)
            self.items.pop()
        else:
            i = self.indexes[val].pop()
            # swap i, n-1
            tail = self.items[n-1]
            self.items[i] = tail
            self.indexes[tail].remove(n-1)
            self.indexes[tail].add(i)
            # remove from items
            self.items.pop()
        if len(self.indexes[val]) == 0:
            del self.indexes[val]

        return True

    def getRandom(self) -> int:
        """
        Get a random element from the collection.
        """
        return random.choice(self.items)


# Your RandomizedCollection object will be instantiated and called as such:
# obj = RandomizedCollection()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
# @lc code=end


class Test(unittest.TestCase):
    def test(self):
        cmds = ["RandomizedCollection", "insert", "remove", "insert", "remove", "getRandom", "getRandom",
                "getRandom", "getRandom", "getRandom", "getRandom", "getRandom", "getRandom", "getRandom", "getRandom"]
        vals = [[], [0], [0], [-1], [0], [],
                [], [], [], [], [], [], [], [], []]

        s = RandomizedCollection()
        for i in range(1, len(cmds)):
            cmd = cmds[i]
            {
                'insert': lambda v: s.insert(v),
                'remove': lambda v: s.remove(v),
                'getRandom': lambda: s.getRandom()
            }[cmd](*vals[i])


if __name__ == "__main__":
    unittest.main()

File: src/381.o-1-时间插入、删除和获取随机元素-允许重复.rs
/*
 * @lc app=leetcode.cn id=381 lang=rust
 *
 * [381] O(1) 时间插入、删除和获取随机元素 - 允许重复
 */

// @lc code=start
use rand::prelude::*;
use std::collections::HashMap;
use std::collections::HashSet;

#[allow(unused)]
struct RandomizedCollection {
    items: Vec<i32>,
    indexes: HashMap<i32, HashSet<usize>>,
}

#[allow(unused)]
impl RandomizedCollection {
    /// Initialize your data structure here.
    fn new() -> Self {
        Self {
            items: vec![],
            indexes: HashMap::new(),
        }
    }

    /// Inserts a value to the collection. Returns true if the collection did not already contain the specified element.
    fn insert(&mut self, val: i32) -> bool {
        // insert to items
        let i = self.items.len();
        self.items.push(val);

        match self.indexes.get_mut(&val) {
            None => {
                // not exists
                let mut new_indexes = HashSet::new();
                new_indexes.insert(i);
                self.indexes.insert(val, new_indexes);
                true
            }
            Some(indexes) => {
                // exists
                indexes.insert(i);
                false
            }
        }
    }

    /// Removes a value from the collection. Returns true if the collection contained the specified element.
    fn remove(&mut self, val: i32) -> bool {
        match self.indexes.get_mut(&val) {
            None => {
                // not exists
                false
            }
            Some(indexes) => {
                // exists
                // pop a [index] = val
                let &index = indexes.iter().next().unwrap();
                indexes.remove(&index);
                // remove entry if empty
                if (indexes.len() == 0) {
                    self.indexes.remove(&val);
                }
                // remove [index]=val from items
                // move [end] to [index]
                if index != self.items.len() - 1 {
                    // find an item to swap
                    let last_index = self.items.len() - 1;
                    let &val_swap = self.items.last().unwrap();
                    let indexes = self.indexes.get_mut(&val_swap).unwrap();
                    self.items[index] = val_swap;
                    indexes.remove(&last_index);
                    indexes.insert(index);
                }
                self.items.pop();
                true
            }
        }
    }

    /// Get a random element from the collection.
    fn get_random(&self) -> i32 {
        let mut r = thread_rng();
        let index = r.gen_range(0, self.items.len());
        self.items[index]
    }
}
// @lc code=end

#[test]
fn test_solution() {
    let mut obj = RandomizedCollection::new();
    macro_rules! test {
        (insert, $args:expr, $ans:expr) => {
            assert_eq!(obj.insert($args), $ans)
        };
        (remove, $args:expr, $ans:expr) => {
            assert_eq!(obj.remove($args), $ans)
        };
    }
    test!(insert, 1, true);
    test!(insert, 1, false);
    test!(insert, 2, true);
    obj.get_random();
    test!(remove, 1, true);
    test!(remove, 2, true);
    assert_eq!(obj.get_random(), 1);
}
